stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift

```diff
@@ -64,12 +67,38 @@ func posixPipe() -> (readFD: CInt, writeFD: CInt) {
 /// stderr.
 public func spawnChild(args: [String])
   -> (pid: pid_t, stdinFD: CInt, stdoutFD: CInt, stderrFD: CInt) {
+  let childStdout = posixPipe()
+  let childStdin = posixPipe()
+  let childStderr = posixPipe()
+
+#if os(Android)
+  // posix_spawn isn't available on Android. Instead, we fork and exec.
+  let pid = fork()
+  if pid < 0 {
+    preconditionFailure("fork() failed")
+  } else if pid == 0 {
+    // pid of 0 means we are now in the child process.
+    // Capture the output and execute the program.
+    dup2(childStdout.writeFD, STDOUT_FILENO)
+    dup2(childStdin.readFD, STDIN_FILENO)
+    dup2(childStderr.writeFD, STDERR_FILENO)
+
+    var result: Int32 = 0
+    withArrayOfCStrings([Process.arguments[0]] + args) {
+      result = execve(Process.arguments[0], $0, _getEnviron())
+    }
+
+    if result != 0 {
+      preconditionFailure("execve() failed. errno: \(errno)")
+    }
+    _exit(126)
```

Comment:
this has different semantics to `posix_spawn`. `posix_spawn` will give you a proper `errno` value if the spawn failed, this just exits the child process. I.e. spawning `["/bin/zsh", "-c", "exit 126;"]` will be indistinguishable for the caller, right? It will trigger the `preconditionFailure`, no idea what that actually means.

I'd propose to open a separate pipe, called `child2parent` or so which will be `fcntl`d to `FD_CLOEXEC`. That means if `execve` succeeds, it will be closed automatically and that means the parent will read EOF. If `execve` fails, we just write `execve`'s `errno` to that pipe so the parent can read it and learns what went wrong in the child. So the parent waits until it can read, if it reads EOF (`read` returns 0) it knows, the child spawned successfully, if it reads anything else, it knows that this is `execve`'s `errno` value and can return that to the caller. Does that make sense?

Here some pseudo-C code which will hopefully make clear what I mean:

    [...]
    } else if (pid == 0) {
        int err = fcntl(child2parent[1], F_SETFD, FD_CLOEXEC); /* This fd will be closed on exec */
        int errno_save = errno;
        if (err) {
           preconditionFailure("fcntl failed, errno: %d", errno_save);
           abort();
        }
        execve(...);
        errno_save = errno;
        /* execve returned, this is an error that we need to communicate back to the parent */
        ssize_t suc_write = write(child2parent[1], &errno_save, sizeof(typeof(errno)));
        errno_save = errno;
        if (suc_write > 0 && suc_write < sizeof(typeof(errno)));
            /* implement write retry ;) */
        } else if (suc_write == 0) {
            /* preconditionFailure("EOF?!"); */
            abort();
        } else if (suc_write < 0) {
            preconditionFailure("write failed, errno %d", errno_save);
            abort();
        }
        close(child2parent[1]);
    }
    [...]